File: scripts/analyze.js
#!/usr/bin/env node

/**
 * analyze.js - Repository Pattern Analyzer
 * 
 * Analyzes all prompts in the repository to extract:
 * - Common patterns
 * - Security rules
 * - Tool architectures
 * - Statistics
 */

const fs = require('fs');
const path = require('path');

const ROOT_DIR = path.join(__dirname, '..');
const EXCLUDED_DIRS = ['.git', '.github', 'site', 'assets', 'node_modules', 'scripts', 'metadata'];

class Analyzer {
  constructor() {
    this.tools = [];
    this.patterns = {
      security: [],
      conciseness: [],
      toolUsage: [],
      verification: [],
      parallel: [],
    };
    this.stats = {
      totalTools: 0,
      totalPrompts: 0,
      totalLines: 0,
      avgPromptLength: 0,
      avgToolCount: 0,
    };
  }

  log(message) {
    console.log(message);
  }

  // Get all tool directories
  getToolDirectories() {
    return fs.readdirSync(ROOT_DIR, { withFileTypes: true })
      .filter(dirent => dirent.isDirectory())
      .map(dirent => dirent.name)
      .filter(name => !EXCLUDED_DIRS.includes(name) && !name.startsWith('.'));
  }

  // Analyze a prompt file
  analyzePrompt(toolName, fileName, content) {
    const lines = content.split('\n');
    const lowerContent = content.toLowerCase();

    // Extract patterns
    const analysis = {
      toolName,
      fileName,
      lineCount: lines.length,
      charCount: content.length,
      patterns: {
        hasSecurityRules: /never.*log.*secret|never.*expose|security|malicious/i.test(content),
        hasConciseness: /concise|brief|short|1-3 sentences|no preamble/i.test(content),
        hasToolInstructions: /tool|function|available tools/i.test(content),
        hasVerification: /verify|check|validate|test/i.test(content),
        hasParallel: /parallel|simultaneously|independent/i.test(content),
        hasTodo: /todo|progress|track/i.test(content),
        hasMemory: /memory|remember|persist|context/i.test(content),
        hasSubAgents: /sub-agent|delegate|oracle|reasoning/i.test(content),
      },
    };

    // Count security mentions
    const securityKeywords = ['never log', 'never expose', 'secret', 'api key', 'password', 'security'];
    analysis.securityMentions = securityKeywords.filter(kw => 
      lowerContent.includes(kw)
    ).length;

    return analysis;
  }

  // Analyze tool directory
  analyzeTool(toolName) {
    const toolPath = path.join(ROOT_DIR, toolName);
    const files = fs.readdirSync(toolPath);

    const promptFiles = files.filter(f => 
      (f.toLowerCase().includes('prompt') || f.toLowerCase().includes('system')) && 
      (f.endsWith('.txt') || f.endsWith('.md'))
    );

    const toolFiles = files.filter(f => f.endsWith('.json'));

    const tool = {
      name: toolName,
      promptCount: promptFiles.length,
      hasTools: toolFiles.length > 0,
      hasReadme: files.includes('README.md'),
      prompts: [],
    };

    // Analyze each prompt
    promptFiles.forEach(file => {
      const content = fs.readFileSync(path.join(toolPath, file), 'utf-8');
      const analysis = this.analyzePrompt(toolName, file, content);
      tool.prompts.push(analysis);
    });

    // Count tools if tools.json exists
    if (tool.hasTools) {
      try {
        const toolsContent = fs.readFileSync(path.join(toolPath, toolFiles[0]), 'utf-8');
        const toolsData = JSON.parse(toolsContent);
        tool.toolCount = Array.isArray(toolsData) ? toolsData.length : 
          (toolsData.tools ? toolsData.tools.length : 0);
      } catch (error) {
        tool.toolCount = 0;
      }
    }

    return tool;
  }

  // Calculate statistics
  calculateStats() {
    this.stats.totalTools = this.tools.length;
    this.stats.totalPrompts = this.tools.reduce((sum, t) => sum + t.promptCount, 0);
    this.stats.totalLines = this.tools.reduce((sum, t) => 
      sum + t.prompts.reduce((s, p) => s + p.lineCount, 0), 0
    );

    const allPrompts = this.tools.flatMap(t => t.prompts);
    this.stats.avgPromptLength = allPrompts.length > 0 ?
      Math.round(allPrompts.reduce((sum, p) => sum + p.lineCount, 0) / allPrompts.length) : 0;

    const toolsWithCounts = this.tools.filter(t => t.toolCount);
    this.stats.avgToolCount = toolsWithCounts.length > 0 ?
      Math.round(toolsWithCounts.reduce((sum, t) => sum + t.toolCount, 0) / toolsWithCounts.length) : 0;
  }

  // Extract common patterns
  extractPatterns() {
    const allPrompts = this.tools.flatMap(t => t.prompts);

    // Count pattern occurrences
    const patternCounts = {
      security: allPrompts.filter(p => p.patterns.hasSecurityRules).length,
      conciseness: allPrompts.filter(p => p.patterns.hasConciseness).length,
      tools: allPrompts.filter(p => p.patterns.hasToolInstructions).length,
      verification: allPrompts.filter(p => p.patterns.hasVerification).length,
      parallel: allPrompts.filter(p => p.patterns.hasParallel).length,
      todo: allPrompts.filter(p => p.patterns.hasTodo).length,
      memory: allPrompts.filter(p => p.patterns.hasMemory).length,
      subAgents: allPrompts.filter(p => p.patterns.hasSubAgents).length,
    };

    return patternCounts;
  }

  // Generate report
  generateReport() {
    const patternCounts = this.extractPatterns();
    const totalPrompts = this.stats.totalPrompts;

    const report = {
      summary: {
        totalTools: this.stats.totalTools,
        totalPrompts: this.stats.totalPrompts,
        totalLines: this.stats.totalLines,
        avgPromptLength: this.stats.avgPromptLength,
        avgToolCount: this.stats.avgToolCount,
      },
      patterns: {
        security: {
          count: patternCounts.security,
          percentage: Math.round((patternCounts.security / totalPrompts) * 100),
        },
        conciseness: {
          count: patternCounts.conciseness,
          percentage: Math.round((patternCounts.conciseness / totalPrompts) * 100),
        },
        toolInstructions: {
          count: patternCounts.tools,
          percentage: Math.round((patternCounts.tools / totalPrompts) * 100),
        },
        verification: {
          count: patternCounts.verification,
          percentage: Math.round((patternCounts.verification / totalPrompts) * 100),
        },
        parallelExecution: {
          count: patternCounts.parallel,
          percentage: Math.round((patternCounts.parallel / totalPrompts) * 100),
        },
        todoTracking: {
          count: patternCounts.todo,
          percentage: Math.round((patternCounts.todo / totalPrompts) * 100),
        },
        memorySystem: {
          count: patternCounts.memory,
          percentage: Math.round((patternCounts.memory / totalPrompts) * 100),
        },
        subAgents: {
          count: patternCounts.subAgents,
          percentage: Math.round((patternCounts.subAgents / totalPrompts) * 100),
        },
      },
      topTools: this.tools
        .sort((a, b) => b.prompts.reduce((s, p) => s + p.lineCount, 0) - 
                        a.prompts.reduce((s, p) => s + p.lineCount, 0))
        .slice(0, 10)
        .map(t => ({
          name: t.name,
          prompts: t.promptCount,
          lines: t.prompts.reduce((s, p) => s + p.lineCount, 0),
          hasTools: t.hasTools,
          toolCount: t.toolCount || 0,
        })),
    };

    return report;
  }

  // Print report to console
  printReport(report) {
    console.log('\n' + '='.repeat(60));
    console.log('📊 Repository Analysis Report');
    console.log('='.repeat(60) + '\n');

    console.log('📈 Summary Statistics:');
    console.log(`  Total Tools: ${report.summary.totalTools}`);
    console.log(`  Total Prompts: ${report.summary.totalPrompts}`);
    console.log(`  Total Lines: ${report.summary.totalLines.toLocaleString()}`);
    console.log(`  Avg Prompt Length: ${report.summary.avgPromptLength} lines`);
    console.log(`  Avg Tool Count: ${report.summary.avgToolCount}\n`);

    console.log('🔍 Pattern Analysis:');
    Object.entries(report.patterns).forEach(([name, data]) => {
      const label = name.replace(/([A-Z])/g, ' $1').trim();
      console.log(`  ${label}: ${data.count} (${data.percentage}%)`);
    });

    console.log('\n🏆 Top 10 Tools by Prompt Size:');
    report.topTools.forEach((tool, index) => {
      console.log(`  ${index + 1}. ${tool.name} - ${tool.lines} lines, ${tool.prompts} prompts, ${tool.toolCount} tools`);
    });

    console.log('\n' + '='.repeat(60) + '\n');
  }

  // Save report to file
  saveReport(report) {
    const outputPath = path.join(ROOT_DIR, 'analysis-report.json');
    fs.writeFileSync(outputPath, JSON.stringify(report, null, 2));
    console.log(`📄 Detailed report saved to: analysis-report.json\n`);
  }

  // Main analysis
  run() {
    console.log('\n🔍 Analyzing Repository...\n');

    // Get all tool directories
    const toolDirs = this.getToolDirectories();
    console.log(`Found ${toolDirs.length} tool directories\n`);

    // Analyze each tool
    toolDirs.forEach(toolName => {
      const tool = this.analyzeTool(toolName);
      this.tools.push(tool);
      console.log(`✓ Analyzed ${toolName}`);
    });

    // Calculate statistics
    this.calculateStats();

    // Generate and print report
    const report = this.generateReport();
    this.printReport(report);

    // Save detailed report
    this.saveReport(report);
  }
}

// Run analyzer
if (require.main === module) {
  const analyzer = new Analyzer();
  analyzer.run();
}

module.exports = Analyzer;

File: scripts/analyze.py
#!/usr/bin/env python3
"""
Analyze and generate statistics about AI tools in the repository.
Creates comparison charts and statistics.
"""

import json
import os
from collections import defaultdict

class Colors:
    BLUE = '\033[94m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    CYAN = '\033[96m'
    RESET = '\033[0m'
    BOLD = '\033[1m'

def load_index():
    """Load the metadata index."""
    index_path = "scripts/index.json"
    if not os.path.exists(index_path):
        print("Error: index.json not found. Run 'python scripts/generate_metadata.py' first.")
        exit(1)

    with open(index_path, 'r', encoding='utf-8') as f:
        return json.load(f)

def print_header(text):
    """Print a formatted header."""
    print(f"\n{Colors.BOLD}{Colors.BLUE}{'='*70}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.BLUE}{text.center(70)}{Colors.RESET}")
    print(f"{Colors.BOLD}{Colors.BLUE}{'='*70}{Colors.RESET}\n")

def print_bar_chart(data, title, max_width=50):
    """Print a horizontal bar chart."""
    print(f"{Colors.BOLD}{title}{Colors.RESET}")

    if not data:
        print("  No data available")
        return

    max_value = max(data.values())

    for key, value in sorted(data.items(), key=lambda x: x[1], reverse=True):
        bar_length = int((value / max_value) * max_width) if max_value > 0 else 0
        bar = '█' * bar_length
        print(f"  {key:30} {Colors.GREEN}{bar}{Colors.RESET} {value}")

def analyze_by_category(index):
    """Analyze tools by category."""
    print_bar_chart(index['stats']['by_category'], "Tools by Category")

def analyze_by_company(index):
    """Analyze tools by company."""
    companies = defaultdict(int)
    for tool in index['tools']:
        company = tool.get('company', 'Unknown')
        companies[company] += 1

    print_bar_chart(dict(companies), "Tools by Company")

def analyze_by_type(index):
    """Analyze tools by type."""
    print_bar_chart(index['stats']['by_type'], "Tools by Type")

def analyze_file_sizes(index):
    """Analyze file sizes and line counts."""
    print(f"{Colors.BOLD}File Size Analysis{Colors.RESET}")

    total_size = 0
    largest_files = []

    for tool in index['tools']:
        for file in tool.get('files', []):
            total_size += file['size']
            largest_files.append((file['path'], file['size'], tool['name']))

    # Sort and get top 10
    largest_files.sort(key=lambda x: x[1], reverse=True)

    print(f"\n  Total Size: {total_size / 1024 / 1024:.2f} MB")
    print(f"\n  {Colors.BOLD}Top 10 Largest Files:{Colors.RESET}")
    for path, size, tool_name in largest_files[:10]:
        size_kb = size / 1024
        print(f"    {path:50} {size_kb:8.1f} KB  ({tool_name})")

def analyze_models(index):
    """Analyze AI models used."""
    print(f"{Colors.BOLD}AI Models Analysis{Colors.RESET}\n")

    model_count = defaultdict(int)
    for tool in index['tools']:
        for model in tool.get('models', []):
            model_count[model] += 1

    if model_count:
        print("  Models mentioned in tools:")
        for model, count in sorted(model_count.items(), key=lambda x: x[1], reverse=True):
            print(f"    {model:30} {count} tool(s)")
    else:
        print("  No model information available in metadata")

def analyze_complexity(index):
    """Analyze complexity based on line count and file count."""
    print(f"{Colors.BOLD}Complexity Analysis{Colors.RESET}\n")

    # Sort by total lines
    by_lines = sorted(index['tools'], key=lambda x: x.get('total_lines', 0), reverse=True)

    print(f"  {Colors.BOLD}Most Complex (by line count):{Colors.RESET}")
    for i, tool in enumerate(by_lines[:10], 1):
        lines = tool.get('total_lines', 0)
        files = tool.get('file_count', 0)
        print(f"    {i:2}. {tool['name']:30} {lines:6,} lines, {files:2} files")

    # Sort by file count
    by_files = sorted(index['tools'], key=lambda x: x.get('file_count', 0), reverse=True)

    print(f"\n  {Colors.BOLD}Most Files:{Colors.RESET}")
    for i, tool in enumerate(by_files[:10], 1):
        files = tool.get('file_count', 0)
        print(f"    {i:2}. {tool['name']:30} {files:2} files")

def generate_comparison_table(index):
    """Generate a markdown comparison table."""
    print(f"{Colors.BOLD}Generating Comparison Table (Markdown){Colors.RESET}\n")

    md = "| Tool | Company | Category | Files | Lines | Models |\n"
    md += "|------|---------|----------|-------|-------|--------|\n"

    for tool in sorted(index['tools'], key=lambda x: x['name']):
        name = tool['name']
        company = tool.get('company', 'Unknown')
        category = tool.get('category', 'Unknown')
        files = tool.get('file_count', 0)
        lines = tool.get('total_lines', 0)
        models = ', '.join(tool.get('models', [])[:2]) if tool.get('models') else 'N/A'
        if len(tool.get('models', [])) > 2:
            models += '...'

        md += f"| {name} | {company} | {category} | {files} | {lines:,} | {models} |\n"

    # Save to file
    output_path = "scripts/comparison_table.md"
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(md)

    print(f"  Saved to: {output_path}")

def main():
    """Run all analyses."""
    index = load_index()

    print_header("AI Tools Repository Analysis")

    # Overall stats
    print(f"{Colors.BOLD}Overall Statistics{Colors.RESET}")
    print(f"  Total Tools: {Colors.CYAN}{index['stats']['total_tools']}{Colors.RESET}")
    print(f"  Total Files: {Colors.CYAN}{index['stats']['total_files']}{Colors.RESET}")
    print(f"  Total Lines: {Colors.CYAN}{index['stats']['total_lines']:,}{Colors.RESET}")

    print()
    analyze_by_category(index)
    print()
    analyze_by_company(index)
    print()
    analyze_by_type(index)
    print()
    analyze_models(index)
    print()
    analyze_complexity(index)
    print()
    analyze_file_sizes(index)
    print()
    generate_comparison_table(index)

    print(f"\n{Colors.GREEN}Analysis complete!{Colors.RESET}\n")

if __name__ == "__main__":
    main()
